src/ns.c

```diff
@@ -434,17 +443,21 @@ hsk_ns_onrecv(
     goto done;
   }
 
+  // TLD must pass HNS rules, if not root or synth
+  if (req->tld[0] != 0x00 && !hsk_dns_name_verify(req->tld))
+    goto fail;
+
   // Requesting a lookup.
   if (req->labels > 0) {
     // Check blacklist.
-    if (strcmp(req->tld, "bit") == 0 // Namecoin
-        || strcmp(req->tld, "eth") == 0 // ENS
-        || strcmp(req->tld, "exit") == 0 // Tor
-        || strcmp(req->tld, "gnu") == 0 // GNUnet (GNS)
-        || strcmp(req->tld, "i2p") == 0 // Invisible Internet Project
-        || strcmp(req->tld, "onion") == 0 // Tor
-        || strcmp(req->tld, "tor") == 0 // OnioNS
-        || strcmp(req->tld, "zkey") == 0) { // GNS
+    if (memcmp(req->tld, "\x03""bit", 4) == 0 // Namecoin
+        || memcmp(req->tld, "\x03""eth", 4) == 0 // ENS
+        || memcmp(req->tld, "\x03""exit", 5) == 0 // Tor
+        || memcmp(req->tld, "\x03""gnu", 4) == 0 // GNUnet (GNS)
+        || memcmp(req->tld, "\x03""i2p", 4) == 0 // Invisible Internet Project
+        || memcmp(req->tld, "\x03""onion", 6) == 0 // Tor
+        || memcmp(req->tld, "\x03""tor", 4) == 0 // OnioNS
+        || memcmp(req->tld, "\x03""zkey", 5) == 0) { // GNS
```

Comment:
Should not this instead be:
```
    if (memcmp(req->tld, "\x03""bit", 4) == 0 // Namecoin
        || memcmp(req->tld, "\x03""eth", 4) == 0 // ENS
        || memcmp(req->tld, "\x04""exit", 5) == 0 // Tor
        || memcmp(req->tld, "\x03""gnu", 4) == 0 // GNUnet (GNS)
        || memcmp(req->tld, "\x03""i2p", 4) == 0 // Invisible Internet Project
        || memcmp(req->tld, "\x05""onion", 6) == 0 // Tor
        || memcmp(req->tld, "\x03""tor", 4) == 0 // OnioNS
        || memcmp(req->tld, "\x04""zkey", 5) == 0) { // GNS
```
First byte size does not match the length of the label.

Reply:
:+1: fixed in f5200ab1251950b897e8ef3f71f5a11737aeaa5a